include/seastar/rpc/rpc_types.hh

```diff
@@ -78,8 +78,12 @@ constexpr connection_id invalid_connection_id = connection_id::make_invalid_id()
 
 std::ostream& operator<<(std::ostream&, const connection_id&);
 
+class server;
+
 struct client_info {
     socket_address addr;
+    rpc::server& server;
```

Comment:
why store `rpc::server&` multiple times? now it's in `client_info` and in `connection`.

why we need this reference in `client_info`? strictly speaking, it's not 'info about client', we can just use our own server instance, e.g. `messaging_service::_server`

Reply:
>  but we have 4 of them with connection ids from the same domain.

Exactly why I included the `server&` reference. If there was a single `server&` in messaging service we wouldn't need the `server&` inside `client_info`.

Reply:
> Exactly why I included the `server&` reference. If there was a single `server&` in messaging service we wouldn't need the `server&` inside `client_info`.

ok

> But I don't have access to connection.

But connection has access to `client_info _info`, we can always use that `server&`, no?

Reply:
> But connection has access to client_info _info, we can always use that server&, no?

I don't understand the point you're trying to make

Reply:
it's a private field with 5 usages

Reply:
Fiiiiiiineeeeee

Reply:
I don't understand where you got the "5 usages" from though, have you seen src/rpc/rpc.cc?